iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -571,6 +571,9 @@ static bool compareShapesEqual(ShapedType lhsType, ValueRange lhsDynamicDims,
     // Static shape equivalence means we can fast-path the check.
     return true;
   }
+  if (lhsType.getRank() != rhsType.getRank()) {
```

Comment:
I am not sure this is correct. I hit an issue here too (https://github.com/google/iree/pull/5392/files#diff-131a92b97e4233f143362f6330d8fcd77b3cf6f9aa4a69b68bb6256519597d08R556) . That seems to be a better check to have here?

Reply:
I've been thinking this for a bit while, and I ended up with this. The main reason is that below for-loop assumes they have the same rank. Otherwise `if (lhsType.isDynamicDim(i) != rhsType.isDynamicDim(i))` condition does not make sense to me. Because if lhs has higher rank, `rhsType.isDynamicDim(i)` will be invalid.

Reply:
Ok, I actually couldnt follow the logic of the for loop below that well. But could you add the check for num elements being equal to this as well?

Reply:
Done.

Based on `compareShapesEqual`, I think it is trying to verify if the shapes are identical. The rank should be the same in this case.

Reply:
After thinking a bit while, I think we don't need to modify the check to `lhsType.getNumElements() == rhsType.getNumElements()`.

This is used for `TensorReshapeOp::fold`, and we can only fold the op only if they have the same shape. Otherwise, it will cause verification failed because the uses would use `src` directly, which has different type. e.g., `reshape src from `tensor<6xf32> to tensor<2x3xf32>`. The consumer would expect a `tensor<2x3xf32>` type. We can not fold it in this case.

If they are static, and we would like to verify if they are the same shape. Checking `lhsType == rhsType` is enough and cleaner.